tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/AspNetCoreBlockMiddlewareIntegrationEnd.cs

```diff
@@ -0,0 +1,63 @@
+// <copyright file="AspNetCoreBlockMiddlewareIntegrationEnd.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#if !NETFRAMEWORK
+using System;
+using System.ComponentModel;
+using System.Threading.Tasks;
+using Datadog.Trace.AppSec;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.Configuration;
+using Microsoft.AspNetCore.Http;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AspNetCore
+{
+    /// <summary>
+    /// The ASP.NET Core middleware integration.
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = AssemblyName,
+        TypeName = ApplicationBuilder,
+        MethodName = "Build",
+        ReturnTypeName = "Microsoft.AspNetCore.Http.RequestDelegate",
+        MinimumVersion = Major3,
+        MaximumVersion = Major7,
+        IntegrationName = nameof(IntegrationId.AspNetCore))]
+    [InstrumentMethod(
+        AssemblyName = AssemblyName,
+        TypeName = InternalApplicationBuilder,
+        MethodName = "Build",
+        ReturnTypeName = "Microsoft.AspNetCore.Http.RequestDelegate",
+        MinimumVersion = Major2,
+        MaximumVersion = Major2,
+        IntegrationName = nameof(IntegrationId.AspNetCore))]
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static class AspNetCoreBlockMiddlewareIntegrationEnd
+    {
+        private const string Major2 = "2";
+        private const string Major3 = "3";
+        private const string Major7 = "7";
+        private const string AssemblyName = "Microsoft.AspNetCore.Http";
+
+        private const string ApplicationBuilder = "Microsoft.AspNetCore.Builder.ApplicationBuilder";
+        private const string InternalApplicationBuilder = "Microsoft.AspNetCore.Builder.Internal.ApplicationBuilder";
+
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget>(TTarget instance)
+            where TTarget : IApplicationBuilder
+        {
+            instance.Components.Insert(0, rd => new BlockingMiddleware(rd).Invoke);
+            instance.Components.Add(rd => new BlockingMiddleware(rd, endPipeline: true).Invoke);
```

Comment:
More a thought for future tuning than this PR, as it probably doesn't happen often, but I wonder if we need to check how many middlewares will run, if there's only one we don't need two blocking middleware

Reply:
The chances of having a single middleware in the pipeline are essentially zero-the only way is if you create an application builder and don't add anything to it-so I really wouldn't worry about this personally 😉